dipy/io/stateful_tractogram.py

```diff
@@ -326,25 +357,16 @@ def remove_invalid_streamlines(self):
         indices_to_keep = np.setdiff1d(np.arange(len(self._tractogram)),
                                        np.array(indices_to_remove)).astype(int)
 
-        tmp_streamlines = \
-            itemgetter(*indices_to_keep)(self.get_streamlines_copy())
-        tmp_data_per_point = {}
-        tmp_data_per_streamline = {}
-
-        for key in self._tractogram.data_per_point:
-            tmp_data_per_point[key] = \
-                self._tractogram.data_per_point[key][indices_to_keep]
-
-        for key in self._tractogram.data_per_streamline:
-            tmp_data_per_streamline[key] = \
-                self._tractogram.data_per_streamline[key][indices_to_keep]
+        tmp_streamlines = self.streamlines[indices_to_keep]
+        tmp_data_per_point = self._tractogram.data_per_point[indices_to_keep]
+        tmp_data_per_streamline =\
+         self._tractogram.data_per_streamline[indices_to_keep]
 
-        self._tractogram = Tractogram(tmp_streamlines,
+        self._tractogram = Tractogram(list(tmp_streamlines),
```

Comment:
list -> Streamlines to avoid potential memory issue

Reply:
Actually I encounter a problem recently were the using the ArraySequence was copy the view while keeping the whole underlying ArraySequence data, which is problematic.

Casting to a list was dropping all attribute and just keep streamlines to recreate a clean ArraySequence not using a view, with the right data.

I lost the data, I don't know where it is. But giving only tmp_streamlines was not working. Have you ever encounter this problem and do you think Streamlines() constructor would solve that ?

Reply:
Indeed, really strange.... 

if we go to the compromise again 😄, you can use `tmp_streamlines.data` to get only the data and not the whole `ArraySequence`. 

To answer your question, I think using `Streamlines()` constructor will solve that but I am not 100% sure.

Reply:
Don't know if that's the expected behaviour, but that was my problem (it was counter intuitive for me at first)
The cleanest solution is to force a copy after the slicing.

`from nibabel.streamlines.array_sequence import ArraySequence
from dipy.tracking.streamline import Streamlines
import numpy as np

# 50 streamlines, 6 points per streamlines
streamlines_a = ArraySequence(np.arange(900).reshape((50,6,3)))
streamlines_b = streamlines_a[0:10]
streamlines_c = ArraySequence(streamlines_a[0:10])
streamlines_d = Streamlines(streamlines_a[0:10])
streamlines_e = ArraySequence(list(streamlines_a[0:10]))
streamlines_f = ArraySequence(streamlines_a[0:10].copy())

# 10 streamlines, 6 points per streamlines
print(len(streamlines_a), len(streamlines_b), len(streamlines_c), len(streamlines_d), len(streamlines_e))

# Should have 300 points for original, 60 points for other
print(len(streamlines_a.data), len(streamlines_b.data), len(streamlines_c.data), len(streamlines_d.data), len(streamlines_e.data), len(streamlines_f.data))`

Reply:
interesting... When I look at the `ArraySequence` code , your demo make totally sense and we can improve that but is it the desired behavior? 

I will need the opinion of @MarcCote, @Garyfallidis, and @arokem.

Thanks for pointing that out! Let's see if we need a new issue

Reply:
Your issue seems similar to what was discussed here: https://github.com/nipy/nibabel/issues/729#issuecomment-467959221

I'm still open to suggestion, to make it more intuitive for the user.

For convenience, here's my original answer
"""
That was the intended behavior but I agree it might be unexpected when looking under the hood. The advanced indexing on `ArraySequence` is a lazy operation to avoid creating more memory if we can. If you want to prune the data, you should call [`.copy()`](https://github.com/nipy/nibabel/blob/master/nibabel/streamlines/array_sequence.py#L255-L268) on the `ArraySequence`, like so
```
In [5]: tgram[idx].streamlines.copy().data
Out[5]: 
array([[2, 2, 2],
       [3, 3, 3],
       [4, 4, 4],
       [5, 5, 5],
       [6, 6, 6]])
```
I think we want with that approach (i.e. explicitly calling `.copy()`) so users are aware that operation will be creating additional memory. I'm open to suggestions to make it clearer to the user?
"""

Reply:
If I can add to this discussion: I feel like having the public `data` getter return data that wasn't part of the tractogram view is pretty dangerous (and I'm not even talking about dipy functions that use `_data` directly).

I like the idea of having a view to save memory, but I would expect public methods to return what the object "says" it contains, not more than that because it is sharing memory with a larger ArraySequence.

Reply:
I agree with you, `.data` should be returning the data of the view and nothing more. Unfortunately, I can't see any way of doing that without implicitly creating a copy of the data :-(. Initially, `.data` was private but we made it public so it could be used "legitely" in Dipy's streamlines manipulation (compress, cluster, etc.).

I just discover https://github.com/scikit-hep/awkward-array, maybe ArraySequence should be replaced with that.

Reply:
Interesting, seems like a good replacement for fast streamline operations

Reply:
@frheault two questions for you. 
1) Why do you need `.data` to be not referring to the original data?
2) Why do you need to modify directly `.data` in your code?

I'm thinking of making `.data` "private" because I agree with @ppoulin91 and it was not intended to be used by the users. The proper way for modifying `.data` is by using `.offsets` and `.lengths` (like in https://github.com/nipy/dipy/blob/master/dipy/tracking/streamlinespeed.pyx#L106).

Users can still modify individual streamline if they want, like so
```python
...
streamline = streamlines[4]
streamline += 10
```

Reply:
I think a good old getter and setter function would be needed.
1) When doing slicing, using a view, it is not intuitive that .data is the whole thing. I know it is for memory reason, but like mentioned earlier this error is very easy to do. But it is easily solve by doing copy every times I am getting a view.
2) For example for warping streamlines it is extremely efficient to just get all points at once, throw them at scipy.interpolate and add a something to each point. Or when getting value from a map, so again interpolation.

If you make .data private, but create a getter/setter function for it that deal with it safely and prevent modification of view. I have no problem with that.
I think trying to solve all problem only using attribute is difficult, you can make it private and create a safe function on the side that is meant for specific use case.

Reply:
> Unfortunately, I can't see any way of doing that without implicitly creating a copy of the data :-(

Maybe by using a masked array. Mask is None at the beginning and each call to `__getitem__` just create/update the mask. (ok, we still increase the memory)

In any case, slicing implies jump in memory. Not sure that all streamlines manipulations on DIPY will like this kind of jump. Currently, even if we improve `data`, it is quite dangerous. So, it is not a bad idea to force a copy when we prune the data. Otherwise, we have to review all streamlines operations in DIPY and make sure they align memory before starting any operations.

Reply:
There are a few specific examples where `._data` is accessed in a dangerous way (without checking if `is_view==True`):
- In dipy, e.g. `dipy.tracking.streamline.transform_streamlines()` with `in_place=True`, or `Dpy.write_tracks()`
- In nibabel, `Tractogram` can get an `ArraySequence` view when creating a new instance, but doesn't do any check when calling `apply_affine()` or `to_world()`.

In the end, both users and developers need to be very careful when playing with the internals of `ArraySequence`. I know I personally had trouble with it before I understood what was happening with the view concept.

Reply:
@skoudoro not sure about `MaskedArray`, the mask has to be the same size as the data, so it would always take ~twice the memory. :( 

@ppoulin91 yes I saw how it is used in Tractogram.apply_affine. That's clearly a bug. The transformation should only affect the "viewed" data (e.g. if we only want to apply a transformation on a specific cluster of streamlines) and not the entire referenced data! I'm trying to come up with an efficient fix for it, but I don't think I can :(. My only option is to loop through the "viewed" streamlines one by one and modify them in-place, or make a copy of the streamlines, first.

Reply:
@MarcCote Yeah, I don't think there is any other way apart from looping, because you never know if the view is contiguous in memory or not... :(